test/filters/test_functor_filter.cpp

```diff
@@ -0,0 +1,197 @@
+/*
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Point CLoud Library (PCL) - www.pointclouds.org
+ * Copyright (c) 2020-, Open Perception
+ *
+ * All rights reserved
+ */
+
+#include <pcl/common/generate.h>
+#include <pcl/filters/functor_filter.h>
+#include <pcl/test/gtest.h>
+#include <pcl/point_types.h>
+
+#include <random>
+
+using namespace pcl;
+
+TEST(FunctorFilterTrait, CheckCompatibility)
+{
+  const auto copy_all = [](PointCloud<PointXYZ>, index_t) { return 0; };
+  EXPECT_TRUE((is_functor_for_filter_v<PointXYZ, decltype(copy_all)>));
+
+  const auto ref_all = [](PointCloud<PointXYZ>&, index_t&) { return 0; };
+  EXPECT_FALSE((is_functor_for_filter_v<PointXYZ, decltype(ref_all)>));
+
+  const auto ref_cloud = [](PointCloud<PointXYZ>&, index_t) { return 0; };
+  EXPECT_FALSE((is_functor_for_filter_v<PointXYZ, decltype(ref_cloud)>));
+
+  const auto const_ref_cloud = [](const PointCloud<PointXYZ>&, index_t) { return 0; };
+  EXPECT_TRUE((is_functor_for_filter_v<PointXYZ, decltype(const_ref_cloud)>));
+
+  const auto const_ref_all = [](const PointCloud<PointXYZ>&, const index_t&) {
+    return 0;
+  };
+  EXPECT_TRUE((is_functor_for_filter_v<PointXYZ, decltype(const_ref_all)>));
+}
+
+struct FunctorFilterRandom : public testing::TestWithParam<std::uint32_t> {
+  void
+  SetUp() override
+  {
+    cloud = make_shared<PointCloud<PointXYZ>>();
+
+    std::uint32_t seed = GetParam();
+    common::CloudGenerator<PointXYZ, common::UniformGenerator<float>> generator{
+        {-10., 0., seed}};
+    generator.fill(20, 20, negative_cloud);
+    generator.setParameters({0., 10., seed});
+    generator.fill(10, 10, positive_cloud);
+    *cloud = negative_cloud + positive_cloud;
+  }
+
+  shared_ptr<PointCloud<PointXYZ>> cloud;
+  PointCloud<PointXYZ> out_cloud, negative_cloud, positive_cloud;
+};
+
+TEST_P(FunctorFilterRandom, functioning)
+{
+
+  const auto lambda = [](const PointCloud<PointXYZ>& cloud, index_t idx) {
+    const auto& pt = cloud[idx];
+    return (pt.getArray3fMap() < 0).all();
+  };
+
+  for (const auto& keep_removed : {true, false}) {
+    FunctorFilter<PointXYZ, decltype(lambda)> filter{lambda, keep_removed};
+    filter.setInputCloud(cloud);
+    const auto removed_size = filter.getRemovedIndices()->size();
+
+    filter.setNegative(false);
+    filter.filter(out_cloud);
+
+    EXPECT_EQ(out_cloud.size(), negative_cloud.size());
+    if (keep_removed) {
+      EXPECT_EQ(filter.getRemovedIndices()->size() + out_cloud.size(), cloud->size());
```

Comment:
This the size of removed indices should be equal to `negative_cloud` correct? If yes, add that check as well.